feature/standardize-reader-api

JacksonMaxfield opened the pull request:
**This PR is pointed at branch `release-4.0` meaning: this will not be released to public until release 4.**

## Description

While we are making changes to APIs for `release-4.0` it seems like now would be a good time to stabilize our reader API as well.

So, as much as I would love there to be a `flake8` plugin to do this:
* function names must begin with a verb
* property names must be nouns

Changes:
* `dtype` on `AICSImage` and `Reader` subclasses is now a property (was a function)
* renamed function `size` to `get_size` on Reader subclasses
* added function `get_size_of_dimension` to Reader as many children had already implemented an undocumented version of it
* removed `load_slice` function from `TiffReader` and `OmeTiffReader` (duplicated work now that `dask_data` is available)
* `size_{dimension}` functions on `Reader` subclasses that implemented them is now a property (was a function, additionally now matches the `AICSImage` API which already was a property)
* removed all custom dimension dims setting from base readers

## Pull request recommendations:
- [x] Name your pull request _your-development-type/short-description_. Ex: _feature/read-tiff-files_
- [x] Link to any relevant issue in the PR description. Ex: _Resolves [gh-12], adds tiff file format support_
- [x] Provide relevant tests for your feature or bug fix.
- [x] Provide or update documentation for any feature added by your pull request.

Thanks for contributing!

JacksonMaxfield commented on aicsimageio/readers/reader.py:
```diff
@@ -331,9 +331,39 @@ def get_image_data(
+        # Set the dims
+        self._dims = dims
+
+    def get_size(self, dims: str = Dimensions.DefaultOrder) -> Tuple[int]:
```
This function and `get_size_of_dimension` could be considered for renaming imo.

AetherUnbound replied:
Should this be `get_shape` instead?

toloudis commented on aicsimageio/readers/reader.py:
```diff
@@ -358,6 +388,24 @@ def size(self, dims: str = Dimensions.DefaultOrder) -> Tuple[int]:
+        """
+        dim = dim.upper()
+        if dim in self.dims:
+            return self.dask_data.shape[self.dims.index(dim)]
```
Are we always using dask to get the shape?

JacksonMaxfield replied:
I believe in most cases we are. Certain readers (`TiffReader` and `CziReader`) are able to get shape without loading the image data / without creating the dask array but as a default implementation this works fine imo.

JacksonMaxfield replied:
I think I should change this to be `self.shape` that way however the Reader implements `shape` it will be routed through there.

toloudis replied:
I think I agree with that.  if ometiff reader and czi reader are getting shape without dask_data, then I feel much better.   Maybe there shouldn't even be a default implementation for this?

JacksonMaxfield replied:
For `shape` you mean?

JacksonMaxfield replied:
Resolved for default implementations of shape in various readers + improved `TiffReader` to override default implementation and read properties from metadata

toloudis commented on aicsimageio/readers/tiff_reader.py:
```diff
@@ -178,80 +185,100 @@ def _read_immediate(self) -> np.ndarray:
+    def shape(self) -> Tuple[int]:
+        if self._shape is None:
+            props = self._get_tiff_properties(self._file, self.specific_s_index)
+            self._dims = props.dims
```
is it ever possible that `_shape is None` but _dims or _dtype was previously set?

JacksonMaxfield replied:
GOOD CATCH. `dims` can be set with the new setter.

JacksonMaxfield replied:
After discussion with @toloudis, I have actually decided to remove dimension setting entirely from the base readers. You can still inform the `AICSImage` object of custom dimensions with `AICSImage(..., known_dims="BLAH")` but the _reader_ objects are to some level the raw data (even with our best guessing on certain readers)

heeler commented on aicsimageio/exceptions.py:
```diff
@@ -22,12 +22,7 @@ class InvalidDimensionOrderingError(Exception):
-
-    def __str__(self):
-        return self.message
+    pass
```
Nice cleanup, I didn't realize the base exception would print a message.

heeler commented on aicsimageio/readers/arraylike_reader.py:
```diff
@@ -52,20 +52,6 @@ def _read_immediate(self):
     def dims(self) -> str:
         return self._dims
 
-    @dims.setter
```
Does this mean telling the reader the dims is no longer an option or has that been shifted into another function?

JacksonMaxfield replied:
> removed all custom dimension dims setting from base readers

Yes. Custom dims setting has been removed. I tried to implement that for all readers but it ended up being more complicated.

From resolved conversation:
> After discussion with @toloudis, I have actually decided to remove dimension setting entirely from the base readers. You can still inform the AICSImage object of custom dimensions with AICSImage(..., known_dims="BLAH") but the reader objects are to some level the raw data (even with our best guessing on certain readers)

heeler commented on aicsimageio/readers/czi_reader.py:
```diff
@@ -454,9 +454,6 @@ def dims(self) -> str:
             self._dims = CziFile(self._file).dims
         return self._dims
 
-    def dtype(self) -> np.dtype:
```
👍

heeler commented on aicsimageio/tests/readers/test_arraylike_reader.py:
```diff
@@ -32,42 +32,12 @@ def test_arraylike_reader(arr, expected_shape, expected_dims):
-
-
-@pytest.mark.parametrize(
-    "expected_starting_dims, set_dims, expected_ending_dims",
```
Do we lose coverage by removing these or were they completely redundant?

JacksonMaxfield replied:
The only thing this tested for was custom dims setting which has been removed entirely so the coverage diff should be disregarded in situations like this.

donovanr commented:
@JacksonMaxfield asked me to make some notes about AICSImage api here.  I'm just going to go through anything in
```
sorted({x for x in dir(image) + list(vars(image).keys()) if not x.startswith("_")})
```
and leave notes.  I haven't used most of the dask stuff.

- `client`: didn't use
- `close`: didn't use 
- `cluster`: didn't use 
- `dask_data`: didn't use 
- `data`: fine as is
- `determine_reader`: fine as is
- `dims`: fine, but an ordered string seems kind of fragile, would a named tuple or something that explicitly pairs an integer dimension index with a name be better?  like that it's not a getter fn.
- `get_channel_names`: this should be a noun / property.  also a list seems fragile, a named tuple or dict that explicitly pairs channel indices to channels names seems better.
- `get_image_dask_data`: these accessors are kind of weird but kind of fine. i noticed that when called with no args its way slower than the `.data` property -- may want to default to that in that case?
- `get_image_data`: these accessors are kind of weird but kind of fine
- `get_physical_pixel_size`: should be a noun / property
- `metadata`: sort of fine. maybe a hint on how to display it to a non expert would be useful.
- `reader`: fine
- `shape`: fine.  maybe should be a named tuple?
- `size`: should not exist / should be changed to match what `numpy.ndarray.size` means, ie `x.size = np.prod(x.shape)`
- `size_c`: should not exist / clutter
- `size_s`: should not exist / clutter
- `size_t`: should not exist / clutter
- `size_x`: should not exist / clutter
- `size_y`: should not exist / clutter
- `size_z`: should not exist / clutter
- `view_napari`: didn't work for me, make install more transparent?

JacksonMaxfield commented:
Thank you so much @donovanr. Some responses.

>     * `client`: didn't use

Removed in 4.0

>     * `close`: didn't use

Removed in 4.0

>     * `cluster`: didn't use

Removed in 4.0

>     * `dask_data`: didn't use

Kept and untouched

>     * `data`: fine as is

Kept and untouched

>     * `determine_reader`: fine as is

Kept and untouched

>     * `dims`: fine, but an ordered string seems kind of fragile, would a named tuple or something that explicitly pairs an integer dimension index with a name be better?  like that it's not a getter fn.

Interesting. Would love thoughts on this from @heeler @AetherUnbound @toloudis 

>     * `get_channel_names`: this should be a noun / property.  also a list seems fragile, a named tuple or dict that explicitly pairs channel indices to channels names seems better.

It unfortunately must be a function as it accepts a parameter of `scene`. Scenes, rarely, but can have different channel names. `get_channel_names` just defaults to `scene=0`

>     * `get_image_dask_data`: these accessors are kind of weird but kind of fine. i noticed that when called with no args its way slower than the `.data` property -- may want to default to that in that case?

Yes without a supporting cluster our single threaded performance is slower. Our `.data` performance is comparable to other readers (minus a current bug that @heeler is working on resolving). See [performance considerations for more details](https://github.com/AllenCellModeling/aicsimageio#performance-considerations)

>     * `get_image_data`: these accessors are kind of weird but kind of fine

Following up from above, what seems odd about these?

>     * `get_physical_pixel_size`: should be a noun / property

Similar to `get_channel_names` this also accepts a `scene` parameter. I remember you also said you would like this to return in the same order as the dims (currently it returns them as `XYZ` but the dims natural order is `ZYX`) is that true still or nah?

>     * `metadata`: sort of fine. maybe a hint on how to display it to a non expert would be useful.

I assume you were using an OME-Tiff or CZI and are referring to the XML. Yea.... That's tricky. I could include something in the docstring but `metadata` for use is like a "you get what you get" because each format is different.

>     * `reader`: fine

Kept and untouched

>     * `shape`: fine.  maybe should be a named tuple?

I think I like your suggestion of `dims` being a `NamedTuple` but I think I would push back that this should match the `numpy` / `dask` API and `shape` should always be a basic `tuple`.

>     * `size`: should not exist / should be changed to match what `numpy.ndarray.size` means, ie `x.size = np.prod(x.shape)`

_Interesting_. I think that's fine as long as we go with the `NamedTuple` approach for dims. More feedback from @heeler @AetherUnbound and @toloudis would be good here.
 
>     * `size_c`: should not exist / clutter
> 
>     * `size_s`: should not exist / clutter
> 
>     * `size_t`: should not exist / clutter
> 
>     * `size_x`: should not exist / clutter
> 
>     * `size_y`: should not exist / clutter
> 
>     * `size_z`: should not exist / clutter

Grouping these together. I agree. I don't think @heeler or @toloudis would though

>     * `view_napari`: didn't work for me, make install more transparent?

Removed in 4.0

donovanr commented:
to be clear those were my very opiniony opinions and mostly i'm just grateful we have this library at all.

re named tuples / keeping shape in sync with numpy, as far as i understand it, named tuples can also just be treated like regular tuples.  so say i did the following
```
Shape = namedtuple("Shape", " ".join(d for d in image.dims))
shape = Shape(*image.shape)
```
that would give me
```
shape
>>> Shape(S=1, T=1, C=6, Z=36, Y=148, X=124)
```
and i can access `shape.Z` and get `36`,  but i can also access it with integer indexing e.g. `shape[3]`, or iterate through it like a regular tuple, or explicitly cast it to a regular tuple with `tuple(shape)` and just work with `(1, 1, 6, 36, 148, 124)` if I really want to.

python experts please let me know if i'm misunderstanding, but are there any real drawbacks?  it seems like a pretty clear case of "Explicit is better than implicit".

toloudis commented:
> to be clear those were my very opiniony opinions and mostly i'm just grateful we have this library at all.
> 
> re named tuples / keeping shape in sync with numpy, as far as i understand it, named tuples can also just be treated like regular tuples. so say i did the following
> 
> ```
> Shape = namedtuple("Shape", " ".join(d for d in image.dims))
> shape = Shape(*image.shape)
> ```
> 
> that would give me
> 
> ```
> shape
> >>> Shape(S=1, T=1, C=6, Z=36, Y=148, X=124)
> ```
> 
> and i can access `shape.Z` and get `36`, but i can also access it with integer indexing e.g. `shape[3]`, or iterate through it like a regular tuple, or explicitly cast it to a regular tuple with `tuple(shape)` and just work with `(1, 1, 6, 36, 148, 124)` if I really want to.
> 
> python experts please let me know if i'm misunderstanding, but are there any real drawbacks? it seems like a pretty clear case of "Explicit is better than implicit".

I guess I slightly prefer returning a `shape` that looks exactly like a numpy shape, only for the reason that it is totally familiar and compatible with numpy.     I kind of like the namedtuple thing too but I would call it something different, like dimshape or dimensions or some hybrid thing.  Can you get an ordered list of the labels in a named tuple?  I have no problem with aicsimageio supporting all these different variants if we can name the api in a reasonable way.

JacksonMaxfield commented:
Closing until #145 is merged, will reopen once those changes make their way into this branch
